File: backend/models/hostel.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type HostelBlock struct {
	ID        string    `gorm:"primaryKey" json:"id"`
	TenantID  string    `gorm:"not null;index" json:"tenant_id"`
	Name      string    `gorm:"not null" json:"name"` // Block A, Boys Hostel etc
	Type      string    `gorm:"not null" json:"type"` // boys, girls, mixed
	Floors    int       `gorm:"not null" json:"floors"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

func (h *HostelBlock) BeforeCreate(tx *gorm.DB) error {
	if h.ID == "" {
		h.ID = uuid.NewString()
	}
	return nil
}

type HostelRoom struct {
	ID         string      `gorm:"primaryKey" json:"id"`
	TenantID   string      `gorm:"not null;index;uniqueIndex:idx_room_block_no" json:"tenant_id"`
	BlockID    string      `gorm:"not null;index;uniqueIndex:idx_room_block_no" json:"block_id"`
	Block      HostelBlock `gorm:"foreignKey:BlockID" json:"block"`
	RoomNumber string      `gorm:"not null;uniqueIndex:idx_room_block_no" json:"room_number"`
	Floor      int         `gorm:"not null" json:"floor"`
	Capacity   int         `gorm:"not null" json:"capacity"`
	Occupied   int         `gorm:"default:0" json:"occupied"`
	RoomType   string      `gorm:"not null" json:"room_type"`         // single, double, triple, dormitory
	Status     string      `gorm:"default:'available'" json:"status"` // available, full, maintenance
	MonthlyFee float64     `gorm:"not null" json:"monthly_fee"`
	// Class-based pricing. RoomClassID links the room to a reusable RoomClass
	// that carries a rate; RateType/RateAmount are an optional per-room override
	// (RateType == "" means "inherit the class rate, or fall back to MonthlyFee").
	RoomClassID *string    `gorm:"index" json:"room_class_id"`
	RoomClass   *RoomClass `gorm:"foreignKey:RoomClassID" json:"room_class,omitempty"`
	RateType    string     `json:"rate_type"`   // override: "", monthly, semester, annual
	RateAmount  float64    `json:"rate_amount"` // override amount, used when RateType != ""
	CreatedAt   time.Time  `json:"created_at"`
	UpdatedAt   time.Time  `json:"updated_at"`
}

func (r *HostelRoom) BeforeCreate(tx *gorm.DB) error {
	if r.ID == "" {
		r.ID = uuid.NewString()
	}
	return nil
}

type HostelAllocation struct {
	ID         string     `gorm:"primaryKey" json:"id"`
	TenantID   string     `gorm:"not null;index" json:"tenant_id"`
	StudentID  string     `gorm:"not null;index" json:"student_id"`
	Student    Student    `gorm:"foreignKey:StudentID" json:"student"`
	RoomID     string     `gorm:"not null;index" json:"room_id"`
	Room       HostelRoom `gorm:"foreignKey:RoomID" json:"room"`
	BedNumber  int        `gorm:"default:0" json:"bed_number"` // 1-based bed slot within the room
	AllocDate  time.Time  `gorm:"not null" json:"alloc_date"`
	VacateDate *time.Time `json:"vacate_date"`
	Status     string     `gorm:"default:'active'" json:"status"` // active, vacated
	CreatedAt  time.Time  `json:"created_at"`
	UpdatedAt  time.Time  `json:"updated_at"`
}

func (a *HostelAllocation) BeforeCreate(tx *gorm.DB) error {
	if a.ID == "" {
		a.ID = uuid.NewString()
	}
	return nil
}

// RoomClass is a reusable pricing/category template for hostel rooms (for
// example "3-Sharing AC" or "Free Sharing"). Rooms link to a class and inherit
// its rate unless they set a per-room override. One rate per class; the
// annual <-> semester split uses the current academic year's semester count,
// computed at read time.
type RoomClass struct {
	ID          string    `gorm:"primaryKey" json:"id"`
	TenantID    string    `gorm:"not null;index" json:"tenant_id"`
	Name        string    `gorm:"not null" json:"name"`
	Description string    `json:"description"`
	RateType    string    `gorm:"not null;default:'semester'" json:"rate_type"` // monthly, semester, annual
	RateAmount  float64   `gorm:"not null;default:0" json:"rate_amount"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

func (rc *RoomClass) BeforeCreate(tx *gorm.DB) error {
	if rc.ID == "" {
		rc.ID = uuid.NewString()
	}
	return nil
}

File: frontend/graphql/mutations/hostel.ts
import { gql } from "@apollo/client";

export const CREATE_HOSTEL_BLOCK = gql`
  mutation CreateHostelBlock($input: CreateHostelBlockInput!) {
    createHostelBlock(input: $input) {
      id name type floors
    }
  }
`;

export const BULK_DELETE_HOSTEL_BLOCKS = gql`
  mutation BulkDeleteHostelBlocks($ids: [ID!]!) {
    bulkDeleteHostelBlocks(ids: $ids)
  }
`;

export const CREATE_HOSTEL_ROOM = gql`
  mutation CreateHostelRoom($input: CreateHostelRoomInput!) {
    createHostelRoom(input: $input) {
      id blockId roomNumber floor capacity occupied roomType status monthlyFee
      roomClassId effectiveRateType semesterRate annualRate monthlyRate
      roomClass { id name }
      block { id name }
    }
  }
`;

export const UPDATE_HOSTEL_ROOM = gql`
  mutation UpdateHostelRoom($id: ID!, $input: UpdateHostelRoomInput!) {
    updateHostelRoom(id: $id, input: $input) {
      id blockId roomNumber floor capacity occupied roomType status monthlyFee
      roomClassId effectiveRateType semesterRate annualRate monthlyRate
      roomClass { id name }
      block { id name }
    }
  }
`;

export const ALLOCATE_HOSTEL_ROOM = gql`
  mutation AllocateHostelRoom($input: AllocateHostelRoomInput!) {
    allocateHostelRoom(input: $input) {
      id studentId roomId bedNumber allocDate vacateDate status
      student { id rollNumber gender user { id name } }
      room { id roomNumber block { id name type } }
    }
  }
`;

export const VACATE_HOSTEL_ROOM = gql`
  mutation VacateHostelRoom($id: ID!, $input: VacateHostelRoomInput!) {
    vacateHostelRoom(id: $id, input: $input) {
      id studentId roomId allocDate vacateDate status
    }
  }
`;

export const CREATE_ROOM_CLASS = gql`
  mutation CreateRoomClass($input: CreateRoomClassInput!) {
    createRoomClass(input: $input) {
      id name description rateType rateAmount semesterRate annualRate monthlyRate
    }
  }
`;

export const UPDATE_ROOM_CLASS = gql`
  mutation UpdateRoomClass($id: ID!, $input: UpdateRoomClassInput!) {
    updateRoomClass(id: $id, input: $input) {
      id name description rateType rateAmount semesterRate annualRate monthlyRate
    }
  }
`;

export const DELETE_ROOM_CLASS = gql`
  mutation DeleteRoomClass($id: ID!) {
    deleteRoomClass(id: $id)
  }
`;

export const BULK_DELETE_HOSTEL_ROOMS = gql`
  mutation BulkDeleteHostelRooms($ids: [ID!]!) {
    bulkDeleteHostelRooms(ids: $ids)
  }
`;

export const BULK_DELETE_HOSTEL_ALLOCATIONS = gql`
  mutation BulkDeleteHostelAllocations($ids: [ID!]!) {
    bulkDeleteHostelAllocations(ids: $ids)
  }
`;

export const BULK_DELETE_ROOM_CLASSES = gql`
  mutation BulkDeleteRoomClasses($ids: [ID!]!) {
    bulkDeleteRoomClasses(ids: $ids)
  }
`;
